mailbox/api/src/test/java/org/apache/james/mailbox/events/EventDeadLettersContract.java

```diff
@@ -0,0 +1,405 @@
+
+            assertThatThrownBy(() -> eventDeadLetters.remove(NULL_GROUP, EVENT_ID_1))
+                .isInstanceOf(IllegalArgumentException.class);
+        }
+
+        @Test
+        default void removeShouldThrowWhenEventIdIsNull() {
+            EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+            assertThatThrownBy(() -> eventDeadLetters.remove(GROUP_A, NULL_EVENT_ID))
+                .isInstanceOf(IllegalArgumentException.class);
+        }
+
+        @Test
+        default void removeShouldThrowWhenBothGroupAndEventIdAreNull() {
+            EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+            assertThatThrownBy(() -> eventDeadLetters.remove(NULL_GROUP, NULL_EVENT_ID))
+                .isInstanceOf(IllegalArgumentException.class);
+        }
+
+        @Test
+        default void removeShouldRemoveMatched() {
+            EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+            eventDeadLetters.store(GROUP_A, EVENT_1).block();
+            eventDeadLetters.store(GROUP_A, EVENT_2).block();
+
+            eventDeadLetters.remove(GROUP_A, EVENT_1.getEventId()).block();
+
+            assertThat(eventDeadLetters.failedEventId(GROUP_A, EVENT_1.getEventId()).block())
+                .isNull();
+        }
+
+        @Test
+        default void removeShouldKeepNonMatched() {
+            EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+            eventDeadLetters.store(GROUP_A, EVENT_1).block();
+            eventDeadLetters.store(GROUP_A, EVENT_2).block();
+            eventDeadLetters.store(GROUP_A, EVENT_3).block();
+
+            eventDeadLetters.remove(GROUP_A, EVENT_1.getEventId()).block();
+
+            assertThat(eventDeadLetters.failedEventIds(GROUP_A).toStream())
+                .containsOnly(EVENT_ID_2, EVENT_ID_3);
+        }
+
+        @Test
+        default void removeShouldNotThrowWhenNoMatched() {
+            EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+            eventDeadLetters.store(GROUP_A, EVENT_1).block();
+
+            assertThatCode(() -> eventDeadLetters.remove(GROUP_A, EVENT_2.getEventId()).block())
+                .doesNotThrowAnyException();
+        }
+
+        @Test
+        default void removeShouldKeepConsistencyWhenConcurrentRemove() throws Exception {
+            EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+            ConcurrentHashMap<Integer, Group> groups = concurrentGroups();
+            ConcurrentHashMap<Integer, Event.EventId> storedEventIds = new ConcurrentHashMap<>();
+            OPERATION_INDEXES.forEach(operationIndex -> {
+                int groupNumber = operationIndex / OPERATION_COUNT;
+                Event.EventId eventId = Event.EventId.random();
+                storedEventIds.put(operationIndex, eventId);
+                eventDeadLetters.store(groups.get(groupNumber), event(eventId));
+            });
+
+            runConcurrent((threadNumber, step) -> {
```

Comment:
Why are we hiding a nice builder? To understand this test I need to jump to `runConcurrent`.

Can we remove that method?